hpx/parallel/algorithms/is_heap.hpp

```diff
@@ -0,0 +1,201 @@
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#if !defined(HPX_PARALLEL_ALGORITHM_IS_HEAP_JUN_22_2017_1705PM)
+#define HPX_PARALLEL_ALGORITHM_IS_HEAP_JUN_22_2017_1705PM
+
+#include <hpx/config.hpp>
+#include <hpx/async.hpp>
+#include <hpx/lcos/dataflow.hpp>
+#include <hpx/lcos/future.hpp>
+#include <hpx/traits/concepts.hpp>
+#include <hpx/traits/is_callable.hpp>
+#include <hpx/traits/is_iterator.hpp>
+#include <hpx/util/invoke.hpp>
+
+#include <hpx/parallel/algorithms/detail/dispatch.hpp>
+#include <hpx/parallel/config/inline_namespace.hpp>
+#include <hpx/parallel/execution_policy.hpp>
+#include <hpx/parallel/executors/execution.hpp>
+#include <hpx/parallel/traits/projected.hpp>
+#include <hpx/parallel/util/detail/algorithm_result.hpp>
+#include <hpx/parallel/util/detail/handle_local_exceptions.hpp>
+#include <hpx/parallel/util/invoke_projected.hpp>
+#include <hpx/parallel/util/projection_identity.hpp>
+#include <hpx/parallel/util/loop.hpp>
+#include <hpx/parallel/util/partitioner.hpp>
+
+#include <boost/exception_ptr.hpp>
+
+#include <algorithm>
+#include <cstddef>
+#include <iterator>
+#include <list>
+#include <type_traits>
+#include <utility>
+
+namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v1)
+{
+    ///////////////////////////////////////////////////////////////////////////
+    // is_heap_until
+    namespace detail
+    {
+        /// \cond NOINTERNAL
+        struct is_heap_until_helper
+        {
+            template <typename ExPolicy, typename RandIter, typename Comp>
+            typename util::detail::algorithm_result<ExPolicy, RandIter>::type
+            operator()(ExPolicy && policy, RandIter first, RandIter last, Comp comp)
+            {
+                typedef util::detail::algorithm_result<ExPolicy, RandIter> result;
+                typedef typename std::iterator_traits<RandIter>::value_type type;
+                typedef typename std::iterator_traits<RandIter>::difference_type
+                    difference_type;
+
+                difference_type count = std::distance(first, last);
+                if (count <= 1)
+                    return result::get(std::move(last));
+
+                RandIter second = first + 1;
+                --count;
+                util::cancellation_token<std::size_t> tok(count);
+
+                return util::partitioner<ExPolicy, RandIter, void>::
+                    call_with_index(
+                        std::forward<ExPolicy>(policy), second, count, 1,
+                        [tok, first, comp](RandIter it, std::size_t part_size,
+                            std::size_t base_idx) mutable
+                        {
+                            util::loop_idx_n(
+                                base_idx, it, part_size, tok,
+                                [&tok, first, &comp](type const& v, std::size_t i)
+                                {
+                                    if (comp(*(first + i / 2), v))
```

Comment:
Since the comparison function is a function object, passing by reference here is fine. Passing by value requires a copy operation on the FO which will be more costly - in cases like this, where a lambda captures the FO (by reference) the compiler only has to call the function at the address supplied - it does not have to 'dereference' the pointer every time, so this is a zero cost reference capture in effect.